### ISSUE TITLE
Why the e2_tts module loss and  DurationPredictor loss are added to a single loss?

### ISSUE THREAD

**Issue by JohnHerry** [bfs18/e2_tts#15]
hi, I need some help about the loss in the project, the reconstruction loss and the duration prediction loss are added as one single loss for optimize, instead of optimize them independently.  since e2_tts and duration_predictor are two different network, why we can make such a design here?

**Comment by bfs18** [comment]
Hi @JohnHerry , Actually, I didn't strictly follow the E2E-TTS paper here. In the current implementation, the duration predictor and acoustic model share the same text encoder, so I optimize their losses together.

**Comment by JohnHerry** [comment]
Thank you for the help.    
My configuration is     
```
backbone:
    class_path: rfwave.dit.DiTRFE2ETTSMutiTaskBackbone
    init_args:
        input_channels: 512
        output_channels: 80
        dim: 1024
        intermediate_dim: 2048
        num_layers: 20
        num_heads: 16

input_adaptor:
    class_path: rfwave.input.E2ECtxCharInputAdaptor
    init_args:
        embedding_dim: 512
        vocab_size: 338
        ctx_dim: 80
        num_layers: 4
```       
My corpus is Chinese/English mixed, about 80K hours totally , ,and the totally phoneme set contains 337 items.  model trained on 6 cards with batch size=10 samples, each sample is about 3 to 20 seconds.   I have trained to about 400K steps, nearly one epoch (every epoch contains more then 500K steps), but the validation output is not good.  
The grad_norm is 0.8 now,   The traning loss is nearly 0.05,  about 0.02 for dur_loss and 0.02 for rf_loss;   but the validation loss is over 0.5.   both training and validation losses refuse to go down significantly.

![A](https://github.com/user-attachments/assets/bcfc82bc-c440-4123-a8dd-83d50708bfaa)
![B](https://github.com/user-attachments/assets/5170d957-aa63-437e-a461-6f6826b5211f)
![D](https://github.com/user-attachments/assets/558ce0c8-5aee-4dd6-9b65-56c6f4123845)


With the vocos vocoder, the generated mel sounds like a puff of nonsense concatenate a puff of noise, no clear speech generated. is there any traning config I can adjust here?

**Comment by JohnHerry** [comment]
another question is about the duration predictor.  if I had configged a largger model like above, is that means I shoud config deeper DurationPredictor model? and a bigger `standalone_dur_start_step` value?

**Comment by JohnHerry** [comment]
请问一下， text encoder的输出，传入duration predcitor的时候，是否有必要detach一下？ duration predictor的梯度回传，会对E2TTS声学模型训练产生负面影响吗？

**Comment by bfs18** [comment]
> 请问一下， text encoder的输出，传入duration predcitor的时候，是否有必要detach一下？ duration predictor的梯度回传，会对E2TTS声学模型训练产生负面影响吗？

影响不大

**Comment by bfs18** [comment]
> another question is about the duration predictor. if I had configged a largger model like above, is that means I shoud config deeper DurationPredictor model? and a bigger `standalone_dur_start_step` value?

No, the duration model's capacity is already redundant, there's no need to increase it.

**Comment by bfs18** [comment]
> Thank you for the help. My configuration is
> 
> ```
> backbone:
>     class_path: rfwave.dit.DiTRFE2ETTSMutiTaskBackbone
>     init_args:
>         input_channels: 512
>         output_channels: 80
>         dim: 1024
>         intermediate_dim: 2048
>         num_layers: 20
>         num_heads: 16
> 
> input_adaptor:
>     class_path: rfwave.input.E2ECtxCharInputAdaptor
>     init_args:
>         embedding_dim: 512
>         vocab_size: 338
>         ctx_dim: 80
>         num_layers: 4
> ```
> 
> My corpus is Chinese/English mixed, about 80K hours totally , ,and the totally phoneme set contains 337 items. model trained on 6 cards with batch size=10 samples, each sample is about 3 to 20 seconds. I have trained to about 400K steps, nearly one epoch (every epoch contains more then 500K steps), but the validation output is not good. The grad_norm is 0.8 now, The traning loss is nearly 0.05, about 0.02 for dur_loss and 0.02 for rf_loss; but the validation loss is over 0.5. both training and validation losses refuse to go down significantly.
> 
> ![A](https://private-user-images.githubusercontent.com/8011802/393672167-bcfc82bc-c440-4123-a8dd-83d50708bfaa.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ3MDYzMzYsIm5iZiI6MTczNDcwNjAzNiwicGF0aCI6Ii84MDExODAyLzM5MzY3MjE2Ny1iY2ZjODJiYy1jNDQwLTQxMjMtYThkZC04M2Q1MDcwOGJmYWEuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMjBUMTQ0NzE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmNkMDY0MTM1MmNkZWM0Yzk0MDMwZWI5NDcxMTY0MTdlNjRlY2NmNmMxOWJiMzlhYThkZWMyNjIzNmNlZTMyZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.YhZFfvKpnU3H0lmpp5m38wSkCU_9RAf96mE_kv6Sx0o) ![B](https://private-user-images.githubusercontent.com/8011802/393672288-5170d957-aa63-437e-a461-6f6826b5211f.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ3MDYzMzYsIm5iZiI6MTczNDcwNjAzNiwicGF0aCI6Ii84MDExODAyLzM5MzY3MjI4OC01MTcwZDk1Ny1hYTYzLTQzN2UtYTQ2MS02ZjY4MjZiNTIxMWYuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMjBUMTQ0NzE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDE4MmQwZjJjNjE0YWUyMzdhMDNiMzc1MTk2ZDFiMmFhZWQ4OTlkNWM3ZDg3OThmMTRlMjQzZjZjYmE1NDg3NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.M8a159E6zh65pLGZQQAe3OrzMrHX4Pl58yTh3VBQEMA) ![D](https://private-user-images.githubusercontent.com/8011802/393672410-558ce0c8-5aee-4dd6-9b65-56c6f4123845.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ3MDYzMzYsIm5iZiI6MTczNDcwNjAzNiwicGF0aCI6Ii84MDExODAyLzM5MzY3MjQxMC01NThjZTBjOC01YWVlLTRkZDYtOWI2NS01NmM2ZjQxMjM4NDUuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMjBUMTQ0NzE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDZiNmNkMmExN2Q3MzA0OTdjM2RmZjhkNGE5NmIyYzFmNDczYjU2Y2E2NTFlYmQzNjM3MzMxMTA3ODFiZmJmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.xyRxO87i3WSMmtR5GEGV_bZ0EO45FlOEOn2rup8H17M)
> 
> With the vocos vocoder, the generated mel sounds like a puff of nonsense concatenate a puff of noise, no clear speech generated. is there any traning config I can adjust here?

你现在应该已经有结果了吧

**Comment by JohnHerry** [comment]
是的，现在有结果了。做了几次不同尝试，开始以为出来的有效合成语音结果是因为自己调小了grad_clip, 来尽可能降低有噪数据影响。后来发现可能还是因为使用了您的bucketed dataloader的缘故。应该是模型要稳定训练，尽量让每个batch的学习语音长度是稳定的。而我们的数据集样本长短差距太大，采用固定数量的batch_size 而不考虑实际的语音内容长度的话，训练起来就不稳定。感谢!
